File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        Set<Character> charSet = new HashSet<>();
        int maxLength = 0;
        int left = 0;
        for(int right=0; right<n; right++) {
            if(!charSet.contains(s.charAt(right))) {
                charSet.add(s.charAt(right));
                maxLength = Math.max(maxLength, right-left+1);
            }
            else {
                while(charSet.contains(s.charAt(right))) {
                    charSet.remove(s.charAt(left));
                    left++;
                }
                charSet.add(s.charAt(right));
            }
        }
        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let map = {};
    let i = j = 0;
    let currWS = 0;
    let maxWS = 0;

    while(i <= j && j < s.length) {
        //if character is already exists in the map
        if(map[s[j]] !== undefined && map[s[j]] >= i) {
            i = map[s[j]] + 1;
            map[s[j]] = j;
        }
        //if character is not in the map
        else{
            map[s[j]] = j;
            currWS = j-i+1;
        }
        j++;
        maxWS = Math.max(currWS, maxWS);
    }
    return maxWS;
};